cmd/ctl/pkg/convert/convert.go

```diff
@@ -0,0 +1,254 @@
+	if err != nil {
+		return err
+	}
+
+	return o.Printer.PrintObj(objects, o.Out)
+}
+
+// asVersionedObject converts a list of infos into a single object - either a List containing
+// the objects as children, or if only a single Object is present, as that object. The provided
+// version will be preferred as the conversion target, but the Object's mapping version will be
+// used if that version is not present.
+func asVersionedObject(infos []*resource.Info, forceList bool, specifiedOutputVersion schema.GroupVersion, encoder runtime.Encoder) (runtime.Object, error) {
+	objects, err := asVersionedObjects(infos, specifiedOutputVersion, encoder)
+	if err != nil {
+		return nil, err
+	}
+
+	var object runtime.Object
+	if len(objects) == 1 && !forceList {
+		object = objects[0]
+	} else {
+		object = &metainternalversion.List{Items: objects}
+
+		targetVersions := []schema.GroupVersion{}
+		if !specifiedOutputVersion.Empty() {
+			targetVersions = append(targetVersions, specifiedOutputVersion)
+		}
+		targetVersions = append(targetVersions, schema.GroupVersion{Group: "", Version: "v1"})
+
+		converted, err := tryConvert(object, targetVersions...)
+		if err != nil {
+			return nil, err
+		}
+
+		object = converted
+	}
+
+	actualVersion := object.GetObjectKind().GroupVersionKind()
+
+	if actualVersion.Version != specifiedOutputVersion.Version {
+		defaultVersionInfo := ""
+		if len(actualVersion.Version) > 0 {
+			defaultVersionInfo = fmt.Sprintf("Defaulting to %q", actualVersion.Version)
+		}
+		klog.V(1).Infof("info: the output version specified is invalid. %s\n", defaultVersionInfo)
+	}
+
+	return object, nil
+}
+
+// asVersionedObjects converts a list of infos into versioned objects. The provided
+// version will be preferred as the conversion target, but the Object's mapping version will be
+// used if that version is not present.
+func asVersionedObjects(infos []*resource.Info, specifiedOutputVersion schema.GroupVersion, encoder runtime.Encoder) ([]runtime.Object, error) {
+	objects := []runtime.Object{}
+	for _, info := range infos {
+		if info.Object == nil {
+			continue
+		}
+
+		targetVersions := []schema.GroupVersion{}
+		// objects that are not part of api.Scheme must be converted to JSON
+		// TODO: convert to map[string]interface{}, attach to runtime.Unknown?
+		if !specifiedOutputVersion.Empty() {
+			if _, _, err := api.Scheme.ObjectKinds(info.Object); runtime.IsNotRegisteredError(err) {
+				// TODO: ideally this would encode to version, but we don't expose multiple codecs here.
+				data, err := runtime.Encode(encoder, info.Object)
+				if err != nil {
+					return nil, err
+				}
+				// TODO: Set ContentEncoding and ContentType.
+				objects = append(objects, &runtime.Unknown{Raw: data})
+				continue
+			}
+
+			targetVersions = append(targetVersions, specifiedOutputVersion)
+		} else {
+			gvks, _, err := api.Scheme.ObjectKinds(info.Object)
```

Comment:
If the err *isn't* nil, should we log a warning to the user? What cases will this occur in? A resource being parsed that is not present in the scheme?

Reply:
It will error here if there is a resource which doesn't have a kind or version set. Do you want to log and error here or hard fail?